jpeg.go

```diff
@@ -5,49 +5,156 @@ import (
 	"encoding/binary"
 	"errors"
 	"io"
+	"io/ioutil"
 )
 
 var headerBytes []byte = []byte{0xff, 0xd8}
 
+// scanReader reads until the scan terminating bytes are encountered
+type scanReader struct {
+	reader *io.Reader
+}
+
+func firstMarker(p []byte) int {
+	markerIndex := -1
+	// Look for an 0xff followed by anything other than 0x00 or [0xd0-0xd7] (those are RST markers)
+	for i := 0x01; i <= 0xff; i++ {
+		if i >= 0xd0 && i <= 0xd7 {
+			continue
+		}
+		markerIndex = bytes.Index(p, []byte{0xff, byte(i)})
+		if markerIndex >= 0 {
+			return markerIndex
+		}
+	}
+	return markerIndex
+}
+
+func (sr *scanReader) Read(p []byte) (n int, err error) {
+	buf := new(bytes.Buffer)
+	teedReader := io.TeeReader(*sr.reader, buf)
```

Comment:
Yeah, looking at it more, seems like a [bufio.Scanner](https://golang.org/pkg/bufio/#Scanner) would really suit this use case.  Maybe worth looking into, but no worries if you prefer the code as-is.

Reply:
Yeah I think the problem w/ the bufio pieces is that I'm not sure there are any guarantees about what happens to the reader it wraps? When this reader hits EOF, it needs to be able to guarantee that the underlying reader is at the precise location where the next byte in the stream should be and I don't *_think_* the bufio tools can do that?

Does that make sense (I honestly have stopped being able to tell re: this reader work :upside_down_face: )?

Reply:
Also, because we're not really tokenizing, I _think_ that ultimately the `bufio.Scanner` would read basically the entire image into memory at once....

Reply:
With `bufio.Reader`, you should have equally direct control of where the underlying reader ends up.  You just gain the ability to peek bytes which AFAICT is essentially what you're doing here.  Like you could peek the next `len(p)` bytes to determine where the next marker is, then read that many bytes into `p`.  A `bufio.Scanner` could be even more straightforward since you just define the split function.  In both cases, you are able to define the buffer size yourself.

However, I say all of this with the _huge_ caveat that I haven't put a whole lot of effort into thinking these alternatives through all the way.  There may be edge cases that are not handled easily or something like that.  Again, all good with me if you prefer what you've got.